pkg/csi/service/wcp/controller_helper.go

```diff
@@ -440,3 +443,88 @@ func getOverlappingNodes(accessibleNodes []string, topologyRequirement *csi.Topo
 	}
 	return overlappingNodes, nil
 }
+
+// canFakeAttach checks if the volume is eligible to be fake attached and returns a bool value
+func canFakeAttach(ctx context.Context, volumeID string, volumeManager cnsvolume.Manager) bool {
+	log := logger.GetLogger(ctx)
+	//Check volume annotations
+	pvcAnn, err := commonco.ContainerOrchestratorUtility.GetVolumeAnnotations(ctx, volumeID)
+	if err != nil {
+		log.Errorf("failed to get volume annotations")
+		return false
+	}
+
+	if val, found := pvcAnn[common.AnnIgnoreInaccessiblePV]; found && val == "yes" {
+		log.Infof("Found %s annotation for volume %s", common.AnnIgnoreInaccessiblePV, volumeID)
+		//Check if volume is inaccessible
+		log.Debugf("Checking volume health on CNS volume: %s", volumeID)
+		vol, err := common.QueryVolumeByID(ctx, volumeManager, volumeID)
+		if err != nil {
+			log.Errorf("failed to query CNS volume")
+			return false
+		}
+
+		if vol.HealthStatus != string(pbmtypes.PbmHealthStatusForEntityUnknown) {
+			volHealthStatus, err := convertVolumeHealthStatus(vol.HealthStatus)
+			if err != nil {
+				log.Errorf("invalid health status: %s for volume: %s", vol.HealthStatus, vol.VolumeId.Id)
+				return false
+			}
+			log.Infof("CNS volume health is: %s", volHealthStatus)
+
+			annotations := make(map[string]string)
+			// If CNS volume health doesn't match with volume health annotation on pvc,
+			// mark volume health on pvc to be updated.
+			if val, found := pvcAnn[common.AnnVolumeHealth]; !found || val != volHealthStatus {
+				annotations[common.AnnVolumeHealth] = volHealthStatus
+			}
+			// If volume is inaccessible, it can be fake attached.
+			// Mark fake-attached annotation on pvc to be updated.
+			if volHealthStatus == common.VolHealthStatusInaccessible {
+				annotations[common.AnnFakeAttached] = "yes"
+			}
+			// Update annotations
+			if err := commonco.ContainerOrchestratorUtility.UpdateVolumeAnnotations(ctx, volumeID, annotations); err != nil {
+				// If updating pvc annotation fails, the volume should not be fake attached
+				return false
+			}
+
+			// If volume is inaccessible, it can be fake attached.
+			return volHealthStatus == common.VolHealthStatusInaccessible
+		}
+		// For all other cases, return false
+		return false
+	}
+	// Annotation is not found or not set to true, return false
+	log.Infof("Annotation %s not found or not set to true for volume %s", common.AnnIgnoreInaccessiblePV, volumeID)
```

Comment:
Instead of logging `or`, can we log the exact condition it fails.
And, shouldn't this is be a warning message ?

Reply:
We're just concerned with if this annotation is set to yes.
If it's not present or not set to yes, they can be treated as same scenario. Hence it's ok to have a single log message for both the conditions IMO.

Also, logging this as info as mostly this annotation will not be present on pvc. So it's not a deviation from normal behavior that it should be logged as warning.